src/data/finset/noncomm_prod.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+
+import data.finset.fold
+
+/-!
+# Products (respectively, sums) over a finset or a multiset.
+
+The regular `finset.prod` and `multiset.prod` require `[comm_monoid α]`.
+Often, there are collections `s : finset α` where `[monoid α]` and we know,
+in a dependent fashion, that for all the terms `∀ (x ∈ s) (y ∈ s), commute x y`.
+This allows to still have a well-defined product over `s`.
+
+## Main definitions
+
+- `finset.noncomm_prod`, requiring a proof of commutativity of held terms
+- `multiset.noncomm_prod`, requiring a proof of commutativity of held terms
+
+## Implementation details
+
+While `list.prod` is defined via `list.foldl`, `noncomm_prod` is defined via
+`list.foldr` for neater proofs and definitions. By the commutativity assumption,
+the two must be equal.
+
+-/
+
+variables {α β : Type*} (f : α → β → β) (op : α → α → α)
+
+namespace multiset
+
+/-- Fold of a `s : multiset α` with `f : α → β → β`, given a proof that `left_commutative f`
+on all elements `x ∈ s`. -/
+def noncomm_foldr (s : multiset α) (comm : ∀ (x ∈ s) (y ∈ s) b, f x (f y b) = f y (f x b)) (b : β) :
+  β :=
+s.attach.foldr (f ∘ subtype.val) (λ ⟨x, hx⟩ ⟨y, hy⟩, comm x hx y hy) b
+
+@[simp] lemma noncomm_foldr_coe (l : list α) (comm) (b : β) :
+  noncomm_foldr f (l : multiset α) comm b = l.foldr f b :=
+begin
+  simp only [noncomm_foldr, coe_foldr, coe_attach, list.attach],
+  rw ←list.foldr_map,
+  simp [list.map_pmap, list.pmap_eq_map]
+end
+
+@[simp] lemma noncomm_foldr_empty (h) (b : β) :
+  noncomm_foldr f (0 : multiset α) h b = b := rfl
+
+lemma noncomm_foldr_cons (s : multiset α) (a : α) (h) (h') (b : β) :
+  noncomm_foldr f (a ::ₘ s) h b = f a (noncomm_foldr f s h' b) :=
+begin
+  induction s using quotient.induction_on,
+  simp
+end
+
+/-- Fold of a `s : multiset α` with `op : α → α → α`, given a proof that `left_commutative op`
+on all elements `x ∈ s`. -/
+def noncomm_fold (s : multiset α) (comm) (a : α) : α :=
```

Comment:
Why can't you use just `noncomm_foldr` everywhere where you write definition (possibly with a type annotation on `f` if needed)? This doesn't seem like a useful separate definition to have.

Reply:
I give it as a separate definition to mirror the style of `multiset.fold` as opposed to `multiset.foldr`.

Reply:
I give it as a separate definition to mirror the style of `multiset.fold` as opposed to `multiset.foldr`.

Reply:
I think you should remove this definition.
`multiset.fold` is convenient, since its assumptions are different and hence it's defined as `foldr op (left_comm _ hc.comm ha.assoc)`. In your case, the assumptions on the operation are exactly the same.

One thing you could do (that also mirrors `multiset.fold` better) is to replace the assumptions by "commutative and associative on `s`".

Reply:
"associative on `s`" unfortunately isn't enough because the intermediate fold results aren't promised to be in `s`, if I understand correctly. But I implemented something akin to what you asked for.

Reply:
I see. Yeah, this is fine by me.

bors merge